src/index.js

```diff
@@ -0,0 +1,8 @@
+'use strict'
+
+module.exports = {
+  Stop: require('./circuit/hop'),
+  Dialer: require('./dialer'),
+  multicodec: require('./multicodec'),
+  tag: 'Circuit'
+}
```

Comment:
Yeah, I understand what you are trying to achieve, I'm just not a fan of how this 'inject' happens if it happens because in reality I still haven't found how does swarm touch the `listener.js` code here if it never gets exported.

Now I see that you export the Listener through the Dialer and sent it all in: https://github.com/libp2p/js-libp2p-swarm/pull/201/files#diff-875da84119066ec244742e95d1209f0cR99 that is confusing given the pattern used by the other modules.

As I mentioned, tying up Relay as transport forces it to follow everything a transport has, which is just going to be a pain in the long term and confusing right now. I do see the cleverness and understand how you achieved it.

What you identified is that there is a pipeline for 'incomming connections' that needs to be separated from the 'createListener/on('connection')' routine, so that that pipeline can take new conns that come from a transport, relay or any other type of IO that appears in the future.